### PR TITLE
Clarify the difference between ImmutableEntry and Map.entry

### PR DESCRIPTION
**Pull request by idegtiarenko**
This change adds an explanation for custom ImmutableEntry and difference to default `Map.entry`

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/common/util/Maps.java [hunk 1]
```diff
@@ -328,41 +328,4 @@ public static <K, V1, V2> Map<K, V2> transformValues(Map<K, V1> source, Function
         }
         return copy;
     }
-
-    /**
-     * An immutable implementation of {@link Map.Entry}.
-     * @param key key
-     * @param value value
-     */
-    public record ImmutableEntry<KType, VType>(KType key, VType value) implements Map.Entry<KType, VType> {
```
Comment (idegtiarenko) [inline server/src/main/java/org/elasticsearch/common/util/Maps.java:337]: Just realized the default one does not permit null key/values. Was it the point for this class?

Comment (arteam) [reply server/src/main/java/org/elasticsearch/common/util/Maps.java:337]: Yes, I believe that was the intention

https://github.com/elastic/elasticsearch/pull/88815#discussion_r930099929

Comment (idegtiarenko) [reply server/src/main/java/org/elasticsearch/common/util/Maps.java:337]: Updated the javadoc to document this difference.

**Review by arteam — APPROVED** [review]
lgtm, nice to see that documented

**Pull request merged**